olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -401,37 +466,72 @@ CancellationToken OlpClient::OlpClientImpl::CallApi(
     const OlpClient::RequestBodyType& post_body,
     const std::string& content_type,
     const NetworkAsyncCallback& callback) const {
+  if (!settings_.network_request_handler) {
+    callback({static_cast<int>(http::ErrorCode::OFFLINE_ERROR),
+              "Network layer offline or missing."});
+    return CancellationToken();
+  }
+
   auto network_request = CreateRequest(path, method, query_params,
                                        header_params, post_body, content_type);
 
-  const auto& retry_settings = settings_.retry_settings;
+  AddBearer(query_params.empty(), *network_request);
 
-  auto network_settings =
+  PendingUrlRequestPtr request_ptr = nullptr;
+  auto& pending_requests = pending_requests_;
+  const auto& url = network_request->GetUrl();
+  CancellationToken cancellation_token;
+
+  // Only merge same request in case there is no body as a body can alter the
+  // outcome of the request and may not match the response of a request with a
+  // different body.
+  bool merge = (!post_body || post_body->empty());
+  OLP_SDK_LOG_DEBUG_F(kLogTag, "CallApi: url='%s', merge='%s'", url.c_str(),
+                      merge ? "true" : "false");
+
+  if (merge) {
+    request_ptr = (*pending_requests)[url];
+
+    // Add callback and prepare CancellationToken
+    auto call_id = request_ptr->Append(callback);
```

Comment:
What if PendingRequests::OnRequestCompleted will be called in between? 
merge will be True and GetRetryCallback(true) will be called. Then it will try to execute PendingRequests::OnRequestCompleted with request ID that is already not in the map which will possibly cause that request will not be finished properly.